### ISSUE TITLE
Flag --force-kusama doesn't work for subcommands

### ISSUE THREAD

**Issue by cecton** [paritytech/polkadot#933]
I just noticed in the code that we are using the force_kusama field of opt.run when running a subcommand (opt.subcommand). This is not valid because if opt.subcommand is "some", all the parameters are given to the subcommand parser.

```
[0] [09:56:58] ~/r/polkadot HEAD > ./target/debug/polkadot build-spec --force-kusama
error: Found argument '--force-kusama' which wasn't expected, or isn't valid in this context

USAGE:
    polkadot build-spec [FLAGS] [OPTIONS]

For more information try --help
```

This was introduced in 5410791274adc0b54a236ff834f6ab9f22722d74. I suppose I will fix it with my polkadot PR related to https://github.com/paritytech/substrate/pull/5271 but I think there is a key pattern here that we should fix to avoid any further mistakes like this one. I believe this is all due to the fact that RunCmd is not part of the subcommands like the other subcommands. This particularity makes things more complicated and error prone.

**Comment by cecton** [comment]
cc @arkpar @gnunicorn

**Comment by cecton** [comment]
`--enable-authority-discovery` and `--grandpa-pause` are also not working for a different reason but similar

**Comment by cecton** [comment]
I believe the only solution would be to allow global parameters like this:

```
polkadot --enable-authority-discovery build-spec --base-path something --dev
```

@gnunicorn let me know your optinion but I don't think we will have much choice

**Comment by gnunicorn** [comment]
There is not reason to not allow additional global parameters to be injected by something using `sc-cli` – quite the opposite, the idea was to allow each component that encompasses a bigger system itself (e.g. consensus engines) to expose their own global (`structopt`) parameters and subcommands that are then squashed together by the implementation and provided back to the subsystem when it is initialized. That is totally within idea of substrate extensibility.

It becomes problematic though, when these are supposed to inform build-in commands. These must be available to the entire system and known from cli without the subcomponents either way. If those are subsystem specific though, we might want to extend the subcommand to  hand over to any potential subsystem that needs it as well.

But sure, this can be done.

**Comment by bkchr** [comment]
> I believe the only solution would be to allow global parameters like this:
> 
> ```
> polkadot --enable-authority-discovery build-spec --base-path something --dev
> ```
> 
> @gnunicorn let me know your optinion but I don't think we will have much choice

We clearly voted against this and IMHO this just looks weird.

**Comment by cecton** [comment]
If we don't do it then I think the only ways would be:
- use structopt internals again 
- use clap directly in substrate (which shouldn't prevent polkadot or any other cli to use structopt I think)

**Comment by bkchr** [comment]
Not really sure what the actual problem is, but from the description it sounds like we access a field that we should not access? Sounds like a problem in the design of the CLI.

**Comment by bkchr** [comment]
Yeah, clearly a huge flaw in the CLI design: https://github.com/paritytech/polkadot/commit/5410791274adc0b54a236ff834f6ab9f22722d74#diff-aa9e17d15ba39424b200b856a3859a03R32

**Comment by cecton** [comment]
The problem is actually this:
```
		let app = app
			.name(Self::get_executable_name())
			.author(Self::get_author())
			.about(Self::get_description())
			.version(full_version.as_str())
			.settings(&[
				AppSettings::GlobalVersion,
				AppSettings::ArgsNegateSubcommands,
				AppSettings::SubcommandsNegateReqs,
			]);
```

There is no way to declare global fields. If we would use structopt to its full potential than we would have global parameters. But since we are restricting it, it's impossible to do it in pure structopt.

**Comment by bkchr** [comment]
In the old CLI implementation `RunCmd` was part of the `SubCommand` enum. This made it impossible to access `RunCmd` when you are in a different `SubCommand`. This is clearly something we should bring back, maybe that is something you can do with your current pr in Substrate?

**Comment by cecton** [comment]
It would prevents the error prone issue we have here to do so but it won't solve the probleme: --force-kusama is impossible to implement with the current constraints.

(It's also not possible to put back RunCmd as a subcommand because it is not a subcommand. To do so you will need to manipulate the internal of structopt...)

There are only 3 choices possible:
- either we finally allow global parameters (which is the reasonable thing to do)
- either we use structopt internal to hack our way around
- either we use clap directly (investigating that right now)

**Comment by bkchr** [comment]
- We fix structopt to not use internals.

`--force-kusama` doesn't make sense for every sub command and I don't want to have flags for sub commands that don't make any sense at all.

**Comment by cecton** [comment]
If --force-kusama makes only sense in RunCmd then we have no problem at all. But in that PR it was implemented for RunCmd and then for all subcommands in the commits right after so I assume it was the intention to make it work everywhere.

(Also because --force-kusama is a parameter that is used during load_spec which is also used on subcommands)

**Comment by bkchr** [comment]
`force-kusama` is some special command. It should not even exists and should be solved differently.

Stuff like authority discovery and grandpa pause are clearly no global commands.

**Comment by cecton** [comment]
That's good news for me :grin: 

Since --force-kusama doesn't work at the moment anyway. Would it be ok if I just remove the parameter for now (or put it in runcmd only) and we would fix this force-kusama in a better way later?

**Comment by bkchr** [comment]
Put it only in `RunCmd` sounds reasonable.

**Comment by cecton** [comment]
Solved in b5cc86af8cc0005c9c902156a43f688de549d3f9
